### ISSUE TITLE
Unwrapping of null value objects [DATAREST-1175]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-rest#1541]
**[Patrik Mihalcin](https://jira.spring.io/secure/ViewProfile.jspa?name=caparzo)** opened **[DATAREST-1175](https://jira.spring.io/browse/DATAREST-1175?redirect=false)** and commented

Hi..
Let's say I have value object:

```java
@Data
public class Description {
    @Column(name = "DESCRIPTION")
    private final String description;
}
```

and entity:

```java
@Entity
public class Product extends AbstractEntity<Long> {

    @JsonUnwrapped
    private final Description description; 

    public Product (Description description) {
        this.description = description;
    }

    protected Specification() {
        this.description = new Description(null);
    }
}
```

and custom Description serializer:

```java
static class DescriptionSerializer extends StdSerializer<Description> {
    DescriptionSerializer() {
        super(Description.class);
    }

    @Override
    public void serialize(Description value, JsonGenerator jgen, SerializerProvider provider) throws IOException {
        if (value != null) {
            jgen.writeString(value.getDescription());
        } else {
            jgen.writeNull();
        }
    }
}
```

In a database, I have null in `DESCRIPTION` column of `PRODUCT` table

When I put breakpoint here:
https://github.com/spring-projects/spring-data-rest/blob/9b0f0c7101cf6e9b460f560901fe61ae11b69285/spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/RepositoryEntityController.java#L202

and inside:

```java
protected Specification() {
        this.description = new Description(null);
}
```

I can see that JPA/Hibernate/Spring Data invokes this empty constructor and `Description` value object is initialized with {{null}, i.e. Description(null).

When I step forward and `RepositoryInvoker` finds all results using `Pageable`, description value object fields in `Product` entity are set to null instead of `Description(null)`.

Why is that?
Is is possible to have it as I describe above?

The side effect of this is that my custom `DescriptionSerializer`
is never invoked because of how Jackson works internally, `UnwrappingBeanPropertyWriter` doesn't unwrap null value objects.

See here for the reference: https://github.com/FasterXML/jackson-databind/blob/master/src/main/java/com/fasterxml/jackson/databind/ser/impl/UnwrappingBeanPropertyWriter.java#L89

So to sum it up:
null value objects are not unwrapped properly using Jackson and I don't see expected JSON content:

```
{"description" : null}
```

what leads to inconsistent REST API where description is omitted completely.

Frontend clients written in React are sensitive to this behaviour.

---
No further details from [DATAREST-1175](https://jira.spring.io/browse/DATAREST-1175?redirect=false)
